### DISCUSSION TITLE
Negative Jacobian - contact using non-conforming mesh

### DISCUSSION THREAD

**Discussion by prajwalka** [idaholab/moose#28091] [Q&A Meshing]
Hi,
My problem is similar to #16616. I'm trying a simple contact problem with a softer cylinder compressed using rigid cylindrical grips.  I've made three separate cylinders using Cubit and meshed them separately (so the meshes in the contact region are not the same on the two contact surfaces). As a result, depending on the primary and secondary surfaces I am choosing in the contact action block, there is or isn't a negative jacobian error at the beginning of the simulation. 

I've attached the mesh file and two sample input files, one of which works, while the other doesn't. Any help on the same is really appreciated. Thank you so much.

[contact.zip](https://github.com/user-attachments/files/16154786/contact.zip)

**Comment by GiudGiud** [comment]
Possibly tied to #27934
Does the "negative Jacobian" error give a really really high number for the id of the element?
What is the error message?

**Reply by prajwalka** [reply]
Yes, the id of the element is large, while the jacobian value is -1e-16 (PFB the error).

Similar to #27934, I get 0 elements with negative volume, overlapping elements, elements with negative jacobian, and element sides with negative jacobian when I run the mesh diagnostics. I get a lot of non-planar element sides, but I don't think they are connected to this issue. The error is for a "sideset" QUAD4 element on the planar contact surface, which looks fine when I visualize using Paraview.

<img width="894" alt="Screenshot 2024-07-09 at 9 18 16 PM" src="https://github.com/idaholab/moose/assets/43631519/09879b74-2ae9-44a7-bfeb-9198996142ac">

**Reply by GiudGiud** [reply]
>  I get a lot of non-planar element sides, but I don't think they are connected to this issue. 

yes correct. 
@lindsayad fyi
I might have time tomorrow.

**Reply by lindsayad** [reply]
What contact formulation are you using? Node-face or mortar? What is driving the blocks into contact?

**Reply by lindsayad** [reply]
If you are driving the blocks into contact via something like a Dirichlet displacement condition, then this can generate negative jacobians very easily. Imagine forcing boundary nodes to move past the interior nodes and you have a negative Jacobian. The fix in this case is to set your displacement Dirichlet conditions to have `preset = false`

**Reply by prajwalka** [reply]
I'm using the node/face formulation itself for a 3D simulation. Nope, this error is occurring even before the simulation starts (in the mesh generation stage itself). I've tried a few other meshes as well, and this error occurs in most of them. For this mesh, when I change the primary and secondary faces, the simulation runs fine for the first few steps, but leads to distortions thereafter (I thought that maybe I'm using the smaller surface as the primary and the larger surface as the secondary). So, I don't think the Dirichlet BCs are causing this error.

**Comment by GiudGiud** [comment]
I added some code that lets us get past these failures in finding contact nodes
It's not perfect yet, I ll be working with Ben to see what I would have missed

https://github.com/idaholab/moose/pull/28149

**Reply by prajwalka** [reply]
Thanks a lot for looking into this. I'll check it out.
